🐞 - Application error: a client-side exception has occurred

### What happened?

My fresh unsend instance app crashes when clicking the "Add Domain" button and shows the common client-side exception white screen with the message: `"Application error: a client-side exception has occurred"`

Browser console displays the following error:

```
Error: A <Select.Item /> must have a value prop that is not an empty string. This is because the Select value can be set to an empty string to clear the selection and show the placeholder.
```


### Replication Steps

1. Deploy a fresh unsend instance
2. Navigate to /domains 
3. Click the "Add Domain" button

### Self hosted or Cloud?

Self hosted

### What browsers are you seeing the problem on?

Chrome (or chrome based like Brave, Arc, etc)

@KMKoushik recording

![chrome-capture-2024-12-28](https://github.com/user-attachments/assets/c5d70906-e995-46b0-8035-713a55bef0c8)

Error message in browser
`3786-5c16ce142dbf441d.js:1 Error: A <Select.Item /> must have a value prop that is not an empty string. This is because the Select value can be set to an empty string to clear the selection and show the placeholder.`

---

I was using Coolify 1-click implementation, my suspicion is that in the first configuration in the `Add a new SES configuration` Dialog, I left `Region` empty (I thought it would be inferred to the value `us-east-1`) and this will cause problems with the shadcn <Select> because the value cannot be undefined.

It's strange because I tried to do this again in a new instance of unsend and I couldn't replicate the error, it seems like there is a validation in the form and an error is displayed, but I'm sure that in some situation it is possible to create a `SES configuration` without the Region value and this will generate problems in the application as we are seeing here.

I tried quickly but I couldn't replicate it in a local environment.

For friends who still have this problem (@amkayondo and @rokitgg), please send a print of the `/admin` url of your instance, I bet that in the `SES Configurations` table there will be at least one `Region` entry with an empty value.

![image](https://github.com/user-attachments/assets/ab74f6e1-25ce-4b47-8e33-93f91b83a0ff)

--- a/apps/web/src/components/settings/AddSesSettings.tsx
+++ b/apps/web/src/components/settings/AddSesSettings.tsx
@@ -18,9 +18,10 @@ import { Button } from "@usesend/ui/src/button";
 import Spinner from "@usesend/ui/src/spinner";
 import { toast } from "@usesend/ui/src/toaster";
 import { isLocalhost } from "~/utils/client";
+import { sesRegionSchema } from "~/lib/zod/ses-setting-schema";
 
 const FormSchema = z.object({
-  region: z.string(),
+  region: sesRegionSchema,
   usesendUrl: z.string().url(),
   sendRate: z.coerce.number(),
   transactionalQuota: z.coerce.number().min(0).max(100),
@@ -48,14 +49,51 @@ export const AddSesSettings: React.FC<SesSettingsProps> = ({ onSuccess }) => {
 export const AddSesSettingsForm: React.FC<SesSettingsProps> = ({
   onSuccess,
 }) => {
-  const addSesSettings = api.admin.addSesSettings.useMutation();
+  const defaultRegion = api.admin.getDefaultSesRegion.useQuery();
 
+  if (defaultRegion.isLoading) {
+    return (
+      <div className="flex min-h-[500px] items-center justify-center">
+        <Spinner className="h-5 w-5" />
+      </div>
+    );
+  }
+
+  if (!defaultRegion.data) {
+    return (
+      <div className="flex min-h-[500px] flex-col items-center justify-center gap-3 text-center">
+        <p className="text-sm text-muted-foreground" role="alert">
+          Failed to load the default AWS region.
+        </p>
+        <Button
+          type="button"
+          variant="outline"
+          onClick={() => void defaultRegion.refetch()}
+        >
+          Retry
+        </Button>
+      </div>
+    );
+  }
+
+  return (
+    <SesSettingsForm
+      defaultRegion={defaultRegion.data}
+      onSuccess={onSuccess}
+    />
+  );
+};
+
+const SesSettingsForm: React.FC<
+  SesSettingsProps & { defaultRegion: string }
+> = ({ defaultRegion, onSuccess }) => {
+  const addSesSettings = api.admin.addSesSettings.useMutation();
   const utils = api.useUtils();
 
   const form = useForm<z.infer<typeof FormSchema>>({
     resolver: zodResolver(FormSchema),
     defaultValues: {
-      region: "",
+      region: defaultRegion,
       usesendUrl: "",
       sendRate: 1,
       transactionalQuota: 50,
@@ -101,10 +139,12 @@ export const AddSesSettingsForm: React.FC<SesSettingsProps> = ({
   }
 
   const onRegionInputOutOfFocus = async () => {
-    const region = form.getValues("region");
+    const region = sesRegionSchema.safeParse(form.getValues("region"));
 
-    if (region) {
-      const quota = await utils.admin.getQuotaForRegion.fetch({ region });
+    if (region.success) {
+      const quota = await utils.admin.getQuotaForRegion.fetch({
+        region: region.data,
+      });
       form.setValue("sendRate", quota ?? 1);
     }
   };
@@ -113,7 +153,7 @@ export const AddSesSettingsForm: React.FC<SesSettingsProps> = ({
     <Form {...form}>
       <form
         onSubmit={form.handleSubmit(onSubmit)}
-        className=" flex flex-col gap-8 w-full"
+        className=" flex min-h-[500px] flex-col gap-8 w-full"
       >
         <FormField
           control={form.control}

--- a/apps/web/src/components/settings/AddSesSettings.tsx
+++ b/apps/web/src/components/settings/AddSesSettings.tsx
@@ -142,10 +142,19 @@ const SesSettingsForm: React.FC<
     const region = sesRegionSchema.safeParse(form.getValues("region"));
 
     if (region.success) {
-      const quota = await utils.admin.getQuotaForRegion.fetch({
-        region: region.data,
-      });
-      form.setValue("sendRate", quota ?? 1);
+      form.clearErrors("region");
+
+      try {
+        const quota = await utils.admin.getQuotaForRegion.fetch({
+          region: region.data,
+        });
+        form.setValue("sendRate", quota ?? 1);
+      } catch {
+        form.setValue("sendRate", 1);
+        form.setError("region", {
+          message: "Unable to load the SES quota for this region",
+        });
+      }
     }
   };
 

--- a/apps/web/src/env.js
+++ b/apps/web/src/env.js
@@ -1,6 +1,7 @@
 import { EmailStatus } from "@prisma/client";
 import { createEnv } from "@t3-oss/env-nextjs";
 import { z } from "zod";
+import { sesRegionSchema } from "~/lib/zod/ses-setting-schema";
 
 export const env = createEnv({
   /**
@@ -39,7 +40,7 @@ export const env = createEnv({
     GOOGLE_CLIENT_SECRET: z.string().optional(),
     AWS_SES_ENDPOINT: z.string().optional(),
     AWS_SNS_ENDPOINT: z.string().optional(),
-    AWS_DEFAULT_REGION: z.string().default("us-east-1"),
+    AWS_DEFAULT_REGION: sesRegionSchema.default("us-east-1"),
     API_RATE_LIMIT: z
       .string()
       .default("1")

--- a/apps/web/src/env.js
+++ b/apps/web/src/env.js
@@ -1,7 +1,6 @@
 import { EmailStatus } from "@prisma/client";
 import { createEnv } from "@t3-oss/env-nextjs";
 import { z } from "zod";
-import { sesRegionSchema } from "~/lib/zod/ses-setting-schema";
 
 export const env = createEnv({
   /**
@@ -40,7 +39,11 @@ export const env = createEnv({
     GOOGLE_CLIENT_SECRET: z.string().optional(),
     AWS_SES_ENDPOINT: z.string().optional(),
     AWS_SNS_ENDPOINT: z.string().optional(),
-    AWS_DEFAULT_REGION: sesRegionSchema.default("us-east-1"),
+    AWS_DEFAULT_REGION: z
+      .string()
+      .trim()
+      .min(1, "Region is required")
+      .default("us-east-1"),
     API_RATE_LIMIT: z
       .string()
       .default("1")

--- a/apps/web/src/lib/zod/ses-setting-schema.ts
+++ b/apps/web/src/lib/zod/ses-setting-schema.ts
@@ -1,0 +1,7 @@
+import { z } from "zod";
+
+export const sesRegionSchema = z.string().trim().min(1, "Region is required");
+
+export function getValidSesRegions(regions: string[]) {
+  return [...new Set(regions.map((region) => region.trim()).filter(Boolean))];
+}

--- a/apps/web/src/lib/zod/ses-setting-schema.unit.test.ts
+++ b/apps/web/src/lib/zod/ses-setting-schema.unit.test.ts
@@ -1,0 +1,24 @@
+import { describe, expect, it } from "vitest";
+import {
+  getValidSesRegions,
+  sesRegionSchema,
+} from "~/lib/zod/ses-setting-schema";
+
+describe("sesRegionSchema", () => {
+  it("rejects empty regions", () => {
+    expect(sesRegionSchema.safeParse("").success).toBe(false);
+    expect(sesRegionSchema.safeParse("   ").success).toBe(false);
+  });
+
+  it("normalizes valid regions", () => {
+    expect(sesRegionSchema.parse(" us-east-1 ")).toBe("us-east-1");
+  });
+});
+
+describe("getValidSesRegions", () => {
+  it("filters legacy empty regions and removes duplicates", () => {
+    expect(
+      getValidSesRegions(["", "  ", "us-east-1", " us-east-1 ", "eu-west-1"]),
+    ).toEqual(["us-east-1", "eu-west-1"]);
+  });
+});

--- a/apps/web/src/server/api/routers/admin.ts
+++ b/apps/web/src/server/api/routers/admin.ts
@@ -11,6 +11,7 @@ import { logger } from "~/server/logger/log";
 import { UseSend } from "usesend-js";
 import { isCloud } from "~/utils/common";
 import { toPlainHtml } from "~/server/utils/email-content";
+import { sesRegionSchema } from "~/lib/zod/ses-setting-schema";
 
 const waitlistUserSelection = {
   id: true,
@@ -67,10 +68,12 @@ export const adminRouter = createTRPCRouter({
     return SesSettingsService.getAllSettings();
   }),
 
+  getDefaultSesRegion: adminProcedure.query(() => env.AWS_DEFAULT_REGION),
+
   getQuotaForRegion: adminProcedure
     .input(
       z.object({
-        region: z.string(),
+        region: sesRegionSchema,
       }),
     )
     .query(async ({ input }) => {
@@ -81,7 +84,7 @@ export const adminRouter = createTRPCRouter({
   addSesSettings: adminProcedure
     .input(
       z.object({
-        region: z.string(),
+        region: sesRegionSchema,
         usesendUrl: z.string().url(),
         sendRate: z.number(),
         transactionalQuota: z.number(),

--- a/apps/web/src/server/api/routers/domain.ts
+++ b/apps/web/src/server/api/routers/domain.ts
@@ -16,15 +16,19 @@ import {
 } from "~/server/service/domain-service";
 import { sendEmail } from "~/server/service/email-service";
 import { SesSettingsService } from "~/server/service/ses-settings-service";
+import {
+  getValidSesRegions,
+  sesRegionSchema,
+} from "~/lib/zod/ses-setting-schema";
 
 export const domainRouter = createTRPCRouter({
   getAvailableRegions: protectedProcedure.query(async () => {
     const settings = await SesSettingsService.getAllSettings();
-    return settings.map((setting) => setting.region);
+    return getValidSesRegions(settings.map((setting) => setting.region));
   }),
 
   createDomain: teamProcedure
-    .input(z.object({ name: z.string(), region: z.string() }))
+    .input(z.object({ name: z.string(), region: sesRegionSchema }))
     .mutation(async ({ ctx, input }) => {
       return createDomain(
         ctx.team.id,

--- a/apps/web/src/server/service/ses-settings-service.ts
+++ b/apps/web/src/server/service/ses-settings-service.ts
@@ -7,6 +7,7 @@ import { EventType } from "@aws-sdk/client-sesv2";
 import { EmailQueueService } from "./email-queue-service";
 import { smallNanoid } from "../nanoid";
 import { logger } from "../logger/log";
+import { sesRegionSchema } from "~/lib/zod/ses-setting-schema";
 
 const GENERAL_EVENTS: EventType[] = [
   "BOUNCE",
@@ -62,6 +63,8 @@ export class SesSettingsService {
     sendingRateLimit: number;
     transactionalQuota: number;
   }) {
+    region = sesRegionSchema.parse(region);
+
     await this.checkInitialized();
     if (this.cache[region]) {
       throw new Error(`SesSetting for region ${region} already exists`);

--- a/apps/web/src/server/service/ses-settings-service.ts
+++ b/apps/web/src/server/service/ses-settings-service.ts
@@ -28,10 +28,12 @@ export class SesSettingsService {
   public static async getSetting(
     region = env.AWS_DEFAULT_REGION
   ): Promise<SesSetting | null> {
+    const normalizedRegion = sesRegionSchema.parse(region);
+
     await this.checkInitialized();
 
-    if (this.cache[region]) {
-      return this.cache[region] as SesSetting;
+    if (this.cache[normalizedRegion]) {
+      return this.cache[normalizedRegion] as SesSetting;
     }
     return null;
   }

